pkg/controller/logstash/stackmon/sidecar.go

```diff
@@ -0,0 +1,109 @@
+	logstashv1alpha1 "github.com/elastic/cloud-on-k8s/v2/pkg/apis/logstash/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/defaults"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/stackmon"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/stackmon/monitoring"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/volume"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/logstash/network"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/k8s"
+)
+
+const (
+	// cfgHashAnnotation is used to store a hash of the Metricbeat and Filebeat configurations.
+	cfgHashAnnotation = "logstash.k8s.elastic.co/monitoring-config-hash"
+
+	logstashLogsVolumeName = "logstash-logs"
+	logstashLogsMountPath  = "/usr/share/logstash/logs"
+)
+
+func Metricbeat(ctx context.Context, client k8s.Client, logstash logstashv1alpha1.Logstash) (stackmon.BeatSidecar, error) {
+	metricbeat, err := stackmon.NewMetricBeatSidecar(
+		ctx,
+		client,
+		commonv1.LogstashMonitoringAssociationType,
+		&logstash,
+		logstash.Spec.Version,
+		metricbeatConfigTemplate,
+		logstashv1alpha1.Namer,
+		fmt.Sprintf("%s://localhost:%d", "http" /*logstash.Spec.HTTP.Protocol()*/, network.HTTPPort),
+		//TODO: integrate username password with Logstash metrics API
+		"", /* no username for metrics API */
+		"", /* no password for metrics API */
+		false,
+	)
+	if err != nil {
+		return stackmon.BeatSidecar{}, err
+	}
+	return metricbeat, nil
+}
+
+func Filebeat(ctx context.Context, client k8s.Client, logstash logstashv1alpha1.Logstash) (stackmon.BeatSidecar, error) {
+	return stackmon.NewFileBeatSidecar(ctx, client, &logstash, logstash.Spec.Version, filebeatConfig, nil)
+}
+
+// WithMonitoring updates the Logstash Pod template builder to deploy Metricbeat and Filebeat in sidecar containers
+// in the Logstash pod and injects the volumes for the beat configurations and the ES CA certificates.
+func WithMonitoring(ctx context.Context, client k8s.Client, builder *defaults.PodTemplateBuilder, logstash logstashv1alpha1.Logstash) (*defaults.PodTemplateBuilder, error) {
+	isMonitoringReconcilable, err := monitoring.IsReconcilable(&logstash)
+	if err != nil {
+		return nil, err
+	}
+	if !isMonitoringReconcilable {
+		return builder, nil
+	}
+
+	configHash := fnv.New32a()
+	volumes := make([]corev1.Volume, 0)
+
+	if monitoring.IsMetricsDefined(&logstash) {
+		b, err := Metricbeat(ctx, client, logstash)
+		if err != nil {
+			return nil, err
+		}
+
+		volumes = append(volumes, b.Volumes...)
+		builder.WithContainers(b.Container)
+		configHash.Write(b.ConfigHash.Sum(nil))
+	}
+
+	if monitoring.IsLogsDefined(&logstash) {
+		// enable Stack logging to write Logstash logs to disk
+		builder.WithEnv(fileLogStyleEnvVar())
+
+		b, err := Filebeat(ctx, client, logstash)
+		if err != nil {
+			return nil, err
+		}
+
+		// create a logs volume shared between Logstash and Filebeat
+		logsVolume := volume.NewEmptyDirVolume(logstashLogsVolumeName, logstashLogsMountPath)
```

Comment:
We may want to revisit this at a later point, as this creates an ephemeral logs volume that will go away when the pod disappears. When we start adding persistent volumes we might want to consider adding a logs volume. Maybe a comment or a `TODO:` here to remind us.